examples/mtl/mtl.metal

```diff
@@ -0,0 +1,87 @@
+#include <metal_stdlib>
+
+using namespace metal;
+
+#define MAX(x, y) ((x) > (y) ? (x) : (y))
+
+#define QK4_0 32
+#define QR4_0 2
+typedef struct {
+    half    d;              // delta
+    uint8_t qs[QK4_0 / 2];  // nibbles / quants
+} block_q4_0;
+
+static void dequantize_row_q4_0(device const block_q4_0 * x, device float * y, int k) {
+    const int qk = QK4_0;
+
+    assert(k % qk == 0);
+
+    const int nb = k / qk;
+
+    for (int i = 0; i < nb; i++) {
+        const half d = x[i].d;
+
+        for (int j = 0; j < qk/2; ++j) {
+            const int x0 = (x[i].qs[j] & 0x0F) - 8;
+            const int x1 = (x[i].qs[j] >>   4) - 8;
+
+            y[i*qk + j + 0   ] = x0*d;
+            y[i*qk + j + qk/2] = x1*d;
+        }
+    }
+}
+
+// TODO: not needed
+constant int nsoftmax [[function_constant(0)]];
+
+kernel void kernel_add(
+        device const float * src0,
+        device const float * src1,
+        device       float * dst,
+        uint gid[[thread_position_in_grid]]) {
+    dst[gid] = src0[gid] + src1[gid];
+}
+
+kernel void kernel_relu(
+        device const float * src,
+        device       float * dst,
+        uint gid[[thread_position_in_grid]]) {
+    dst[gid] = max(0.0f, src[gid]);
+}
+
+// TODO: broken
+kernel void kernel_soft_max(
+        device const float * src,
+        device       float * dst) {
+    float max = 0.0f;
+    for (int i = 0; i < nsoftmax; i++) {
+        max = MAX(max, src[i]);
+    }
+    float sum = 0.0f;
+    for (int i = 0; i < nsoftmax; i++) {
+        dst[i] = exp(src[i] - max);
+        sum += dst[i];
+    }
+    for (int i = 0; i < nsoftmax; i++) {
+        dst[i] /= sum;
+    }
+}
+
+// TODO: not tested
+kernel void kernel_get_rows_q4_0(
+        device const  void * src0,
+        device const   int * src1,
+        device       float * dst,
+        constant   int64_t & ne00,
+        constant  uint64_t & nb01,
+        constant  uint64_t & nb1,
+        uint gid[[thread_position_in_grid]]) {
+    device const block_q4_0 * src = (device const block_q4_0 *)src0;
+
+    const int i = gid;
+    const int r = ((device int32_t *) src1)[i];
+
+    dequantize_row_q4_0(
+            (device const block_q4_0 *) ((device char *) src0 + r*nb01),
+                       (device float *) ((device char *)  dst + i*nb1), ne00);
+}
```

Comment:
```suggestion
}

// SIMD shuffle instructions require Metal 3 support, although the
// function itself is heavily optimized for the Apple 7 architecture
//
// TODO: Support different quantization formats through function constants.
kernel void gemv_quantized_i4
 (
  device uchar4 *weights [[buffer(0)]],
  device half2 *scales [[buffer(1)]],
  device half *y [[buffer(2)]],
  device half *dst [[buffer(3)]],
  uint tid [[thread_position_in_grid]])
{
  // 8-wide groupings of threads, each thread reads 8 values per iteration.
  // 'groupings of threads' != 'threadgroups'
#define WEIGHTS_PER_UINT 8
#define GROUPING_SIZE 8
  
  uint row = tid / GROUPING_SIZE;
  ushort local_id = tid % GROUPING_SIZE;
  float acc = 0;
  
  // Changing this to a `while` loop harms performance. Perhaps it triggers a
  // separate assembly instruction for control flow.
  for (uint i = 0; i < ncols;) {
    uchar4 vui = weights[i / WEIGHTS_PER_UINT + local_id];
    const uint blocks_in_row = ncols / 32;
    half2 d = scales[row * blocks_in_row / 2 + i / 32 / 2];
    
    {
      half4 y_value = *(device half4*)(y + i + 4 * local_id);
      i += 4 * GROUPING_SIZE;
      
      const short vi0 = vui.x & 0xF;
      const short vi1 = vui.x >> 4;
      float v0 = (vi0 - 8) * d.x;
      float v1 = (vi1 - 8) * d.x;
      acc += v0 * y_value[0];
      acc += v1 * y_value[1];
      
      const short vi2 = vui.y & 0xF;
      const short vi3 = vui.y >> 4;
      float v2 = (vi2 - 8) * d.x;
      float v3 = (vi3 - 8) * d.x;
      acc += v2 * y_value[2];
      acc += v3 * y_value[3];
    }
    
    {
      half4 y_value = *(device half4*)(y + i + 4 * local_id);
      i += 4 * GROUPING_SIZE;
      
      const short vi0 = vui.z & 0xF;
      const short vi1 = vui.z >> 4;
      float v0 = (vi0 - 8) * d.y;
      float v1 = (vi1 - 8) * d.y;
      acc += v0 * y_value[0];
      acc += v1 * y_value[1];
      
      const short vi2 = vui.w & 0xF;
      const short vi3 = vui.w >> 4;
      float v2 = (vi2 - 8) * d.y;
      float v3 = (vi3 - 8) * d.y;
      acc += v2 * y_value[2];
      acc += v3 * y_value[3];
    }
  }
  
  acc += quad_shuffle_xor(acc, 1);
  acc += quad_shuffle_xor(acc, 2);
  acc += simd_shuffle_xor(acc, 4);
  dst[row] = acc;
#undef WEIGHTS_PER_UINT
#undef GROUPING_SIZE
}

// TODO: kernel void gemv_quantized_i5
// TODO: kernel void gemv_quantized_i8
```

Current iteration of the function from my codebase.

Reply:
The same function can be tweaked to support Q4_1. Ideally you use a function constant to decide between Q4_0 and Q4_1. Same for the Int5 and Int8 functions.

Reply:
To use the most optimized one, you need to swizzle how the weights are stored in memory. This format will be incompatible with CPU, but trivially to convert back <-> forth in-place. Look at how the weights are accessed in shader code, and rearrange the data accordingly.

Reply:
> To use the most optimized variant, you need to swizzle how the weights are stored in memory. This format will not be immediately readable by the CPU, but trivial to convert back <-> forth in-place. Look at how the weights are accessed in shader code, and rearrange the data accordingly.

The plan is to try and see what's the best I can get with the current `Q4_0` format, as this is what comes from the exported computation graph. If this is not enough for performance, then the fallback option would be to have a `Q4_0M` metal-specific quantization, but I hope to avoid this.

Reply:
You can probably revert a few of my optimizations to support that format. The format switch occurred either when unrolling 2 iterations or unrolling 4 iterations. There might also be in-between approaches that maximize perf of in-place operation at reduced bandwidth. Check out SIMD-group shuffles (which require a static lane index) and quad-group shuffles (where the fetched lane index can be chosen dynamically).

You build the graph once and then execute many times? If so, you could make a GPU shader that mutates the weights in-place before execution. Then, mutate them into the original format afterward.

Reply:
If you're talking about deinterleaving, that is an absolute must. In fact, you should deinterleave the CPU code right now, to get a massive speedup _without using GPU_. Deinterleaving unlocks a host of memory coalescing optimizations regardless of the processor.

On your M1 Pro, you might be able to saturate bandwidth, making GPU unnecessary.  Metal does have a theoretical advantage for M1/M2 Max and M1/M2 Ultra, where it has double the bandwidth of the CPU. If properly written, I expect relative speeds like this:

| Chip | Current | CPU Speed | GPU Speed |
| ---- | -------- | --------- | --------- |
| M1   | 40 GB/s  | 50 GB/s   | 50 GB/s   |
| M1 Pro | 90 GB/s | 150 GB/s | 150 GB/s |
| M1 Max | 95 GB/s | 150 GB/s | 300 GB/s |
| M1 Ultra | 190 GB/s | 300 GB/s | 600 GB/s |

What Metal does do, is reduce the cost of GEMM and elementwise ops. You can't reach the time theorized by bandwidth of when most of the time is spent on matmul. That's where Accelerate/AMX, FlashAttention, and minimizing driver overhead become critical.

Reply:
A lot of the current wisdom that you refer to assumes F32 Matrix x Matrix multiplication with big sizes.
However, here we are working with 4-bit Matrix x F32 Vector multiplications, so things look different. At least this is my experience so far - could be wrong.

Deinterleaved Q4 layout is something that we can try, but as a fallback option. We have already experimented with it on the CPU and haven't observed actual benefit.

The plan is first to see what is the maximum we can squeeze out of the current format on the GPU, before complicating the implementation with even more layouts.

> You can probably revert a few of my optimizations to support that format

Would you be interested to demonstrate this so we get a feeling of what would be possible in terms of throughput?

Reply:
Almost all attempted optimizations fail, even when there's still a lower minimum somewhere in the design space.

> Would you be interested to demonstrate this so we get a feeling of what would be possible in terms of throughput?

Not sure I have the bandwidth to do this myself. You can follow my comments, and apply them one at a time to your OpenCL shader. You need a high-resolution means to measure bandwidth, such as my [CPU code](https://github.com/ggerganov/llama.cpp/issues/1545#issuecomment-1565820578) in the other thread. Getting this right requires carefully taking many small steps that inch bandwidth higher. I'm not sure the person making the CPU deinterleaved format was this careful, which would explain why it's slower.